service/controller/release/resource/status/cluster.go

```diff
@@ -0,0 +1,212 @@
+	// e.g. operatorVersions["aws-operator"]["8.7.6"]:true
+
+	for _, c := range clusters {
+		fmt.Printf("Cluster %s (%s) is using operator version %s\n", c.ID, c.ReleaseVersion, c.OperatorVersion)
+		releaseVersions[c.ReleaseVersion] = true
+
+		if operatorVersions[c.ProviderOperator] == nil {
+			operatorVersions[c.ProviderOperator] = make(map[string]bool)
+		}
+		operatorVersions[c.ProviderOperator][c.OperatorVersion] = true
+	}
+
+	return releaseVersions, operatorVersions
+}
+
+// Returns a list of tenant clusters currently running on the installation.
+func (r *Resource) getCurrentTenantClusters(ctx context.Context) ([]TenantCluster, error) {
+
+	var tenantClusters []TenantCluster
+	{
+		awsClusters, err := r.getCurrentAWSClusters(ctx)
+		if IsResourceNotFound(err) {
+			// Fall through
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, awsClusters...)
+		r.logger.Log("level", "debug", "message", fmt.Sprintf("found %d aws tenant clusters", len(awsClusters)))
+
+		azureClusters, err := r.getCurrentAzureClusters(ctx)
+		if IsResourceNotFound(err) || IsNoMatchesForKind(err) {
+			// Fall through
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, azureClusters...)
+		r.logger.Log("level", "debug", "message", fmt.Sprintf("found %d azure tenant clusters", len(azureClusters)))
+
+		legacyClusters, err := r.getLegacyClusters(ctx)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, legacyClusters...)
+		r.logger.Log("level", "debug", "message", fmt.Sprintf("found %d legacy tenant clusters", len(legacyClusters)))
+	}
+
+	return tenantClusters, nil
+}
+
+// Returns a list of AWS clusters according to the awscluster resource (non-legacy).
+func (r *Resource) getCurrentAWSClusters(ctx context.Context) ([]TenantCluster, error) {
+	awsclusters, err := r.k8sClient.G8sClient().InfrastructureV1alpha2().AWSClusters("default").List(ctx, metav1.ListOptions{})
+	if IsResourceNotFound(err) {
+		// Fall through
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var clusters []TenantCluster
+	for _, cluster := range awsclusters.Items {
+		c := TenantCluster{
+			ID:               cluster.Name,
+			OperatorVersion:  cluster.Labels[apiexlabels.AWSOperatorVersion],
+			ReleaseVersion:   cluster.Labels[apiexlabels.ReleaseVersion],
+			ProviderOperator: key.ProviderOperatorAWS,
+		}
+		clusters = append(clusters, c)
+	}
+
+	return clusters, nil
+}
+
+// Returns a list of legacy clusters based on <provider>config resources.
+func (r *Resource) getLegacyClusters(ctx context.Context) ([]TenantCluster, error) {
+	var legacyClusters []TenantCluster
+	aws, err := r.getLegacyAWSClusters(ctx)
+	if IsResourceNotFound(err) {
+		// Fall through
+	} else if err != nil {
+		r.logger.Log("level", "error", "message", fmt.Sprintf("error getting aws legacy clusters: %s", err))
```

Comment:
Why only log the error here?

Reply:
Because IMO we should still try to get the other provider types

Reply:
I see, I would have expected any error we want to continue on to go into the fall through. Now we take essentially the same action in both cases of the `if` statement.

Could we then change the log level?

Reply:
Change the log level to what? The `IsResourceNotFound` case is expected 2/3 of the time so I don't think we should log those. 

The intention here was to avoid a case where for example the operator tries to get `awsconfig`s on a KVM installation but it fails for some other reason, and then the operator never gets to the `kvmconfigs`. We should not eat that error, but we also probably don't want to give up.

Reply:
I am not too fussed about this either way - so not blocking here.

My concern is that we will not notice a true issue here since any error is just logged. If the error has no impact then it should be a debug log IMO. I think we should go into a direction where we can return the error here and fail reconciliation but alert / collect those errors (through log collection, metrics, whatever).

Reply:
I see -- I think this could also return the error as I don't recall seeing it since working out some bugs. I will change this and re-test.